File: sources/Android/android-communications/library/src/sdk/java/com/polar/sdk/api/PolarActivityApi.kt
package com.polar.sdk.api

import com.polar.sdk.api.model.activity.Polar247HrSamplesData
import com.polar.sdk.api.model.activity.Polar247PPiSamplesData
import com.polar.sdk.api.model.activity.PolarActiveTimeData
import com.polar.sdk.api.model.activity.PolarCaloriesData
import com.polar.sdk.api.model.activity.PolarDistanceData
import com.polar.sdk.api.model.activity.PolarStepsData
import com.polar.sdk.api.model.sleep.PolarNightlyRechargeData
import com.polar.sdk.impl.utils.CaloriesType
import com.polar.sdk.api.model.activity.PolarActivitySamplesDayData
import com.polar.sdk.api.model.activity.PolarDailySummaryData
import java.time.LocalDate

/**
 * Polar activity API.
 * Requires feature FEATURE_POLAR_ACTIVITY_DATA
 */
interface PolarActivityApi {

    /**
     * Get steps for a given period. Requires feature [PolarBleSdkFeature.FEATURE_POLAR_ACTIVITY_DATA]
     *
     * @param identifier The Polar device ID or BT address.
     * @param fromDate The starting date of the period to retrieve steps from.
     * @param toDate The ending date of the period to retrieve steps from.
     * @return A list of [PolarStepsData] representing the steps data for the specified period.
     */
    suspend fun getSteps(identifier: String, fromDate: LocalDate, toDate: LocalDate): List<PolarStepsData>

    /**
     * Get activity sample data for a given period. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_ACTIVITY_DATA]
     *
     * @param identifier The Polar device ID or BT address.
     * @param fromDate The starting date of the period to retrieve activity sample data from.
     * @param toDate The ending date of the period to retrieve activity sample data from.
     * @return A list of [PolarActivitySamplesDayData] containing all
     * activity sample data for the given date range.
     */
    suspend fun getActivitySampleData(identifier: String, fromDate: LocalDate, toDate: LocalDate): List<PolarActivitySamplesDayData>

    /**
     * Get daily summary sample data for a given period. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_ACTIVITY_DATA] Daily summary data is a cumulative sum for the activity per given date.
     *
     * @param identifier The Polar device ID or BT address.
     * @param fromDate The starting date of the period to retrieve daily summary data from.
     * @param toDate The ending date of the period to retrieve daily summary data from.
     * @return A list of [PolarDailySummaryData] containing all
     * daily summary data for the given date range.
     */
    suspend fun getDailySummaryData(identifier: String, fromDate: LocalDate, toDate: LocalDate): List<PolarDailySummaryData>

    /**
     * Get distance for a given period. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_ACTIVITY_DATA]
     *
     * @param identifier The Polar device ID or BT address.
     * @param fromDate The starting date of the period to retrieve distance from.
     * @param toDate The ending date of the period to retrieve distance from.
     * @return A list of [PolarDistanceData] representing the distance data for the specified period.
     */
    suspend fun getDistance(identifier: String, fromDate: LocalDate, toDate: LocalDate): List<PolarDistanceData>

    /**
     * Get active time for a given period. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_ACTIVITY_DATA]
     *
     * @param identifier The Polar device ID or BT address.
     * @param fromDate The starting date of the period to retrieve active time from.
     * @param toDate The ending date of the period to retrieve active time from.
     * @return A list of [PolarActiveTimeData] representing the active time data for the specified period.
     */
    suspend fun getActiveTime(identifier: String, fromDate: LocalDate, toDate: LocalDate): List<PolarActiveTimeData>

    /**
     * Get specific calories type for a given period. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_ACTIVITY_DATA]
     *
     * @param identifier The Polar device ID or BT address.
     * @param fromDate The starting date of the period to retrieve calories data from.
     * @param toDate The ending date of the period to retrieve calories data from.
     * @param caloriesType The type of calories data to retrieve (e.g., ACTIVITY, TRAINING, BMR).
     * @return A list of [PolarCaloriesData] representing the calories data for the specified period.
     */
    suspend fun getCalories(identifier: String, fromDate: LocalDate, toDate: LocalDate, caloriesType: CaloriesType): List<PolarCaloriesData>

    /**
     * Get 24/7 heart rate samples for a given period. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_ACTIVITY_DATA]
     *
     * @param identifier The Polar device ID or BT address.
     * @param fromDate The starting date of the period to retrieve heart rate samples from.
     * @param toDate The ending date of the period to retrieve heart rate samples from.
     * @return A list of [Polar247HrSamplesData] representing the heart rate samples for the specified period.
     */
    suspend fun get247HrSamples(identifier: String, fromDate: LocalDate, toDate: LocalDate): List<Polar247HrSamplesData>

    /**
     * Get nightly recharge for a given period. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_ACTIVITY_DATA]
     *
     * @param identifier The Polar device ID or BT address.
     * @param fromDate The starting date of the period to retrieve nightly recharge from.
     * @param toDate The ending date of the period to retrieve nightly recharge from.
     * @return A list of [PolarNightlyRechargeData] representing the nightly recharge data for the specified period.
     */
    suspend fun getNightlyRecharge(identifier: String, fromDate: LocalDate, toDate: LocalDate): List<PolarNightlyRechargeData>

    /**
     * Load 24/7 PPi data from a device for a given period. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_ACTIVITY_DATA]
     *
     * @param identifier, Polar device ID or BT address
     * @param fromDate The starting date of the period to retrieve 24/7 PPi data from.
     * @param toDate The ending date of the period to retrieve 24/7 PPi data from.
     * @return A list of [Polar247PPiSamplesData] representing the 24/7 PPi data for the specified period.
     */
    suspend fun get247PPiSamples(identifier: String, fromDate: LocalDate, toDate: LocalDate): List<Polar247PPiSamplesData>
}

File: sources/iOS/ios-communications/Sources/PolarBleSdk/sdk/api/PolarActivityApi.swift
///  Copyright © 2024 Polar. All rights reserved.

import Foundation
import RxSwift

/// Protocol defining methods to get steps for a given period.
public protocol PolarActivityApi {
    /// Get steps for a given period.
    ///
    /// - Parameters:
    ///   - identifier: The Polar device ID or BT address.
    ///   - fromDate: The starting date of the period to retrieve steps from.
    ///   - toDate: The ending date of the period to retrieve steps from.
    /// - Returns: A Single emitting an array of `PolarStepsData` representing the steps data for the specified period.
    func getSteps(identifier: String, fromDate: Date, toDate: Date) -> Single<[PolarStepsData]>
    
    /// Get distance for a given period.
    ///
    /// - Parameters:
    ///   - identifier: The Polar device ID or BT address.
    ///   - fromDate: The starting date of the period to retrieve distance from.
    ///   - toDate: The ending date of the period to retrieve distance from.
    /// - Returns: A Single emitting an array of `PolarDistanceData` representing the distance data for the specified period.
    func getDistance(identifier: String, fromDate: Date, toDate: Date) -> Single<[PolarDistanceData]>

    /// Get active time for a given period.
    ///
    /// - Parameters:
    ///   - identifier: The Polar device ID or BT address.
    ///   - fromDate: The starting date of the period to retrieve active time from.
    ///   - toDate: The ending date of the period to retrieve active time from.
    /// - Returns: A Single emitting an array of `PolarActiveTimeData` representing the active time data for the specified period.
    func getActiveTime(identifier: String, fromDate: Date, toDate: Date) -> Single<[PolarActiveTimeData]>

    /// Get calories for a given period.
    ///
    /// - Parameters:
    ///   - identifier: The Polar device ID or BT address.
    ///   - fromDate: The starting date of the period to retrieve calories from.
    ///   - toDate: The ending date of the period to retrieve calories from.
    ///   - caloriesType The type of calories data to retrieve (e.g., ACTIVITY, TRAINING, BMR).
    /// - Returns: A Single emitting an array of `PolarCaloriesData` representing the calories data for the specified period.
    func getCalories(identifier: String, fromDate: Date, toDate: Date, caloriesType: CaloriesType) -> Single<[PolarCaloriesData]>
    
    /// Get 24/7 heart rate samples for a given period.
    ///
    /// - Parameters:
    ///   - identifier: The Polar device ID or BT address.
    ///   - fromDate: The starting date of the period to retrieve heart rate samples from.
    ///   - toDate: The ending date of the period to retrieve heart rate samples from.
    /// - Returns: A Single emitting an array of `PolarActiveTimeData` representing the heart rate samples data for the specified period.
    func get247HrSamples(identifier: String, fromDate: Date, toDate: Date) -> Single<[Polar247HrSamplesData]>

    /// Get nightly recharge for a given period.
    ///
    /// - Parameters:
    ///   - identifier: The Polar device ID or BT address.
    ///   - fromDate: The starting date of the period to retrieve nightly recharge from.
    ///   - toDate: The ending date of the period to retrieve nightly recharge from.
    /// - Returns: A Single emitting an array of `PolarNightlyRechargeData` representing the nightly recharge data for the specified period.
    func getNightlyRecharge(identifier: String, fromDate: Date, toDate: Date) -> Single<[PolarNightlyRechargeData]>

    /// Get skin temperature for a given period.
    ///
    /// - Parameters:
    ///   - identifier: The Polar device ID or BT address.
    ///   - fromDate: The starting date of the period to retrive skin temperature from.
    ///   - toDate: The ending date of the period to retrieve skin temperature from.
    /// - Returns: A Single emitting an array of `PolarSkinTemperatureResult` representing the skin temperature data for the specified period.
    func getSkinTemperature(identifier: String, fromDate: Date, toDate: Date) -> Single<[PolarSkinTemperatureData.PolarSkinTemperatureResult]>
    
    /// Load 24/7 PPi data from a device for a given period.
    ///
    /// - Parameters:
    ///   - identifier: Polar device ID or BT address
    ///   - fromDate: The starting date of the period to retrieve 24/7 PPi data from
    ///   - toDate: The ending date of the period to retrieve 24/7 PPi data from
    /// - Returns: A [Single] emitting a list of [Polar247PPiSamplesData] representing the 24/7 PPi data for the specified period.
    func get247PPiSamples(identifier: String, fromDate: Date, toDate: Date) -> Single<[Polar247PPiSamplesData]>

    /// Load activity sample data from a device for a given period.
    ///
    /// - Parameters:
    ///   - identifier: Polar device ID or BT address
    ///   - fromDate: The starting date of the period to retrieve activity sample data from
    ///   - toDate: The ending date of the period to retrieve activity sample data from
    /// - Returns: A [Single] emitting a list of [PolarActivityDayData] representing the activity sample data for the specified period.
    func getActivitySampleData(identifier: String, fromDate: Date, toDate: Date) -> Single<[PolarActivityDayData]>
    
    
    /// Load daily summary data from a device for a given period. Daily summary is a cumulative sum for activity per given date.
    ///
    /// - Parameters:
    ///   - identifier: Polar device ID or BT address
    ///   - fromDate: The starting date of the period to retrieve daily summary data from
    ///   - toDate: The ending date of the period to retrieve daily summary data from
    /// - Returns: A [Single] emitting a list of [PolarDailySummary] representing the daily summary data for the specified period.
    func getDailySummaryData(identifier: String, fromDate: Date, toDate: Date) -> Single<[PolarDailySummary]>
}
